plugins/sql.c

```diff
@@ -524,18 +524,41 @@ static int _sqlite3_escape_str(char *to, const char *from)
     return 0;
 }
 
-static int sqlite3_my_callback(void *pArg, int argc __attribute__((unused)),
-			      char **argv,
-			      char **columnNames __attribute__((unused)))
+typedef struct sql_callback_data {
+    char *value;
+    size_t size;
+    size_t *value_len;
+    size_t num_columns;
+    size_t num_values;
+} sql_callback_data_t;
+
+static int sqlite3_my_callback(void *pArg, int num_columns,
+			      char **column_values,
+			      char **column_names __attribute__((unused)))
 {
-    char **result = (char**)pArg;
+    sql_callback_data_t *data = (sql_callback_data_t*)pArg;
+    const char *value = NULL;
 
-    if (argv == NULL) {
-	*result = NULL;				/* no record */
-    } else if (argv[0] == NULL) {
-	*result = strdup(SQL_NULL_VALUE);	/* NULL IS SQL_NULL_VALUE */
+    data->num_columns = num_columns;
+
+    if (column_values == NULL) {
+	value = NULL;			/* no record */
+    } else if (column_values[0] == NULL) {
+	value = SQL_NULL_VALUE;		/* NULL IS SQL_NULL_VALUE */
+	data->num_values++;
     } else {
-	*result = strdup(argv[0]);
+	value = column_values[0];
+	data->num_values++;
+    }
+
+    /* now get the result set value and value_len */
+    /* we only fetch one because we don't care about the rest */
```

Comment:
Getting more than one result should be a failure, certainly it is a security vulnerability to have ambiguous credentials. If an attacker can insert their own creds in front of the valid ones, etc...